lib/iris/experimental/xarray_dataset_wrapper.py

```diff
@@ -0,0 +1,415 @@
+        separately.
+
+        """
+        self._xr = xr
+        self._name = name
+        self._group = group
+
+    @property
+    def name(self):
+        return self._name
+
+    def group(self):
+        return self._group
+
+    def __eq__(self, other):
+        return self._xr == other._xr
+
+    def __ne__(self, other):
+        return not self == other
+
+
+class DimensionMimic(_XrMimic):
+    """
+    A Dimension object mimic wrapper.
+
+    Dimension additional properties: length, unlimited
+
+    NOTE: a DimensionMimic does *NOT* contain an xarray object representing the
+    dimension, because xarray doesn't have such objects.
+    So, in xarray, you can't rename or modify an existing Dataset dimension.
+    But you can re-order, add, and remove ones that no variable uses.
+
+    """
+
+    def __init__(self, name, len, isunlimited=False, group=None):
+        # Note that there *is* no underlying xarray object.
+        # So we make up something, to support equality checks.
+        id_placeholder = (name, len, isunlimited)
+        super().__init__(xr=id_placeholder, name=name, group=group)
+        self._len = len  # A private version, for now, in case needs change.
+        self._unlimited = isunlimited
+
+    @property
+    def size(self):
+        return 0 if self.isunlimited() else self.len
+
+    def __len__(self):
+        return self._len
+
+    def isunlimited(self):
+        return self._unlimited
+
+
+class _Nc4AttrsMimic(_XrMimic):
+    """
+    A class mixin for a Mimic with attribute access.
+
+    I.E. shared by variables and datasets.
+
+    """
+
+    def ncattrs(self):
+        return self._xr.attrs.keys()  # Probably do *not* need/expect a list ?
+
+    def getncattr(self, attr_name):
+        if attr_name in self._xr.attrs:
+            result = self._xr.attrs[attr_name]
+        else:
+            raise AttributeError()
+        return result
+
+    def __getattr__(self, attr_name):
+        return self.getncattr(attr_name)
+
+    #
+    # writing
+    #
+    def setncattr(self, attr_name, value):
+        if isinstance(value, bytes):
+            value = value.decode()
+        self._xr.attrs[attr_name] = value
+
+    # NOTE: not currently supporting ".my_attribute = value" type access.
+    # def __setattr__(self, attr_name, value):
+    #     self.setncattr(attr_name, value)
+
+
+class VariableMimic(_Nc4AttrsMimic):
+    """
+    A Variable object mimic wrapper.
+
+    Variable additional properties:
+        dimensions, dtype, data (+ attributes, parent-group)
+        shape, size, ndim
+
+    """
+
+    @property
+    def dtype(self):
+        return self._xr.dtype
+
+    def chunking(self):
+        return None
+
+    @property
+    def datatype(self):
+        return self.dtype
+
+    @property
+    def dimensions(self):
+        return self._xr.dims
+
+    def __getitem__(self, keys):
+        if self.ndim == 0:
+            return self._xr.data
```

Comment:
`keys` is ununsed in this branch of the if. Is that right? Should `keys` have default value? Document with a comment either way I think.